Update stored file metadata to include a range

cshannon opened the pull request:
This updates the file metadata column in the metadata table so that file references are no longer just a path but also include a range. The field is now serialized as Json. We will need to handle upgrades/backwards compatibility at some point but for now this just changes everything to Json.

The plan is work on some integration tests to test out the functionality for the serialization/storage along with the fenced reader when this is merged.

cshannon committed:
WIP - update stored file metadata to include a range

File metadata is now serialized as Json and contains a range along with
the path for a file

cshannon committed:
Merge branch 'no-chop-merge' into accumulo-1327-store-range

cshannon committed:
Fix table import/export and bulk import

cshannon committed:
Fix MetadataConstraintsTest

cshannon commented:
The sunny IT build is finally passing, I kicked off a full build to see what else is broken.

cshannon committed:
Merge branch 'no-chop-merge' into accumulo-1327-store-range

cshannon committed:
Fix several ITs

cshannon commented:
I stopped the full IT build as I saw several test failures and it was slowing down the build. I went ahead and fixed the ones that popped up so far and re-kicked off another build.

cshannon committed:
fix more ITs and some refactoring

cshannon commented:
Test run showed 8 failures, I fixed them and kicked off another full build, hopefully everything passes now.

cshannon commented:
@keith-turner - The full ITs all passed so this is ready for a review from you to see what you think. I'll leave it as a draft PR for now but can switch it if you think it looks ok but I'm sure there are some things that need fixing. The next step would be to write an IT to test out adding ranges to the metadata table and reading them with the iterator.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/metadata/AbstractTabletFile.java:
```diff
@@ -62,4 +62,17 @@ public Range getRange() {
+  /**
+   * Return the exact string (full Json with range) that is stored in the metadata table.
+   */
+  public abstract String getMetadata();
```
Why were these methods relating to what is in the metadata table added here instead of just in StoredTabletFile?

cshannon replied:
I had added it as ReferencedTabletFile needs metadata for inserts but it doesn't really make sense to put it there as shown by the Unsupported operations in UnreferencedTabletFile. I will remove it and if ReferencedTabletFile is doing an insert it will just create a StoredTabletFile and delegate like it has before.

cshannon commented on server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManager.java:
```diff
@@ -92,7 +93,7 @@ public Path getVolume(Path path) {
 
     public Path removeVolume(Path path) {
-      String pathString = path.toString();
+      String pathString = Objects.requireNonNull(path).toString();
```
Good point, probably just force of habit but it is the same.

cshannon commented on server/base/src/main/java/org/apache/accumulo/server/fs/VolumeUtil.java:
```diff
@@ -162,11 +163,12 @@ public static TabletFiles updateTabletVolumes(ServerContext context, ServiceLock
     }
 
     for (Entry<StoredTabletFile,DataFileValue> entry : tabletFiles.datafiles.entrySet()) {
-      String metaPath = entry.getKey().getMetaUpdateDelete();
```
I figured when we start to use the ranges we will need to add tests for things like this to verify they are properly used. It would obviously be pretty bad if a range got dropped.

keith-turner replied:
Yeah looking at this I was thinking what is the minimal amount of test and code changes that is the most maintainable and ensures we have this correct.  It seems like we may need a lot of tests that verify lots of operations work properly when there is a range present, I am not sure if there is anyway to avoid that.  We may want to open follow on issues for each test that is needed.

cshannon committed:
Update based on comments, remove un-needed methods in abstract class

cshannon committed:
Merge branch 'no-chop-merge' into accumulo-1327-store-range

keith-turner commented:
As part of the elasticity work we created a few tables (like this [one](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=247828052)) and these were really helpful in thinking through things.  I was wondering if it would be useful to create a table for these range changes and was thinking about something like the following.

 Operation | Needs path | Needs range | updates tablet files
 -|-|-|-
Scan | Y | Y | N
Compaction | Y | Y | N
Replace volumes | Y | N | Y
GC ref scan | Y | N | N
Set property | N | N | N

cshannon commented:
> As part of the elasticity work we created a few tables (like this [one](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=247828052)) and these were really helpful in thinking through things. I was wondering if it would be useful to create a table for these range changes and was thinking about something like the following.
> 
> Operation	Needs path	Needs range	updates tablet files
> Scan	Y	Y	N
> Compaction	Y	Y	N
> Replace volumes	Y	N	Y
> GC ref scan	Y	N	N
> Set property	N	N	N

A table like this is really helpful. It should make it much easier to verify correctness and to test if we document what needs to care about certain things. As I start to write tests I can reference this to make sure things look good.

cshannon committed:
Merge branch 'no-chop-merge' into accumulo-1327-store-range

cshannon committed:
Merge branch 'no-chop-merge' into accumulo-1327-store-range

cshannon committed:
Merge branch 'no-chop-merge' into accumulo-1327-store-range

cshannon committed:
MetadataConstraints and test fix after merging main

cshannon committed:
Switch TabletFile range to be exclusive/inclusive to match Tablets (KeyExtent)

cshannon commented:
I've started working on ITs to show that the metadata can be read properly and fencing works. There's a few tests here so far showing contiguous and disjointed ranges work: https://github.com/cshannon/accumulo/blob/accumulo-1327-itest/test/src/main/java/org/apache/accumulo/test/functional/FileMetadataIT.java

I'm going to continue working on the tests, I plan to add tests to show compaction will work and also splits, etc. The test branch is based off of this current PR/branch so will eventually be based off of no-chop-merge branch when this PR is merged in.

After some more tests which should be done soon, I plan to start prototyping dropping the chop compactions and updating merge to use ranges. The majority of this work appears to be in TabletGroupWatcher in the [mergeMetadataRecords](https://github.com/apache/accumulo/blob/ffe0018b513d582840e0cde04a0772f59deaea72/server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java#L715) method.

cshannon committed:
Update compareTo to use path and range object and not json to properly sort

cshannon commented:
I added compaction to the tests and compaction seems to work correctly with the simple tests, here is some output showing before and after: https://gist.github.com/cshannon/67a72d3075c07992eb64a36e64ecbcda

I started testing out splits and split doesn't seem to be correct. When I wrote a test that had fenced rfiles in metadata and then I added splits, the output seems to be not right for the resulting new metadata. So I plan to look into the split code usage for the ranges later this week and also I will test out GC as well.

cshannon committed:
Add missing range to ReferencedTabletFile inside StoredTabletFile constructor

cshannon committed:
Fix json encoding by including entire Key and not just the row portion

This fixes an issue where extra bytes were appended on deserialization
when trying to create keys that were inclusive

cshannon committed:
Merge branch 'no-chop-merge' into accumulo-1327-store-range

keith-turner commented on core/src/main/java/org/apache/accumulo/core/metadata/StoredTabletFile.java:
```diff
@@ -121,11 +150,105 @@ public String toString() {
+        gson.fromJson(Objects.requireNonNull(json), TabletFileCqMetadataGson.class);
+    // If we have previously enforced the exclusive/inclusive of a range then can just set that here
+    return new TabletFileCq(new Path(URI.create(metadata.path)),
+        new Range(decodeRow(metadata.startRow), false, decodeRow(metadata.endRow), true));
```
Could probably avoid encoding the entire Key with the following change.

```suggestion
        new Range(decodeRow(metadata.startRow), true, decodeRow(metadata.endRow), false));
```

The reason being if you look at this [Range constructor](https://github.com/apache/accumulo/blob/0a3feba20f2a0d0d542748776777969ba9dcd9b1/core/src/main/java/org/apache/accumulo/core/data/Range.java#L115-L124) it does a few things.

 * It does not mutate the start row when startRowInclusive is true, it takes it as is.
 * It does not mutate the end row when endRowInclusive is false, it takes it as is.
 * No matter what startRowInclusive and endRowInclusive are set to, the inclusivity of the range is always true, false.

I wrote  the following little unit test to double check this and it passed.

```java
  @Test
  public void testMisc() {
    Range range1 = new Range("abc", false, "def", true);
    Range range2 = new Range(range1.getStartKey().getRow(), true,
                                                   range1.getEndKey().getRow(), false);
    assertEquals(range1, range2);
  }
```

cshannon replied:
Wouldn't this cause potential issues with things like equality? The problem I see is that the original Range is constructed using false/true for the start/end row inclusive and then we would reconstruct with the opposite. So I could see scenarios where we'd end up with Ranges that are equal but equality returns false due to those flags being backwards.

cshannon replied:
And right after I typed my message I saw your unit test that passes equality with the flags switched, let me take a closer look at that as that seems odd to me because I did a glance at the Range class and it seemed like it was using the flags as part of the equality check.

cshannon replied:
Ok, I probably should have read your entire message more closely. :) as you pointed out it will always set the inclusivity of the range is always true, false so I will make this change as this should work, thanks.

keith-turner replied:
> Ok, I probably should have read your entire message more closely. :)

I have done that too many times.  I'd like to think I am getting better and learning from my mistakes, but not sure.

cshannon committed:
Revert "Fix json encoding by including entire Key and not just the row portion"

This reverts commit 80623450c768f40494160c1cafaec8a61bf7311a.

cshannon committed:
Switch decoding Range json to using true/false for inclusivity

This prevents the rows from being modified and this constructor being
used always sets the inclusivity of start/end to be true/false

cshannon committed:
Fix formatting

cshannon committed:
Add validation for Range inclusivity for TabletFiles

keith-turner commented:
@cshannon I had meant to create a PR for 883d3d8b32faf839b98b6f3a604bc0d3f8813302, but accidentally pushed it directly to the no chop branch.  Let me know if you have any feedback for that.  Also I was wondering where you would recommend I put a unit test for StoredTabletFile.

keith-turner commented:
@cshannon I was also wondering if the validation is needed in any other places than where I added it in 883d3d8b32faf839b98b6f3a604bc0d3f8813302

cshannon commented:
> @cshannon I had meant to create a PR for [883d3d8](https://github.com/apache/accumulo/commit/883d3d8b32faf839b98b6f3a604bc0d3f8813302), but accidentally pushed it directly to the no chop branch. Let me know if you have any feedback for that. Also I was wondering where you would recommend I put a unit test for StoredTabletFile.

This looks good but I think we can probably just move the validation to AbstractTabletFile so I can  move it. I had already added some validation but this looks better. For a test, we can probably just created a StoredTabletFile test class that lives next to where ReferencedTabletFileTest is.
